### PR TITLE
Use etcd 2.3.7 (container already pushed)

### LINKED ISSUE #32253: Etcd3 -> etcd 2 rollback: etcd crashes
Noticed fiddling with live cluster. etcd datadir and controller/apiserver logs: 
https://console.cloud.google.com/storage/browser/devnul/etcd-crash

Etcd logs are full of:
```
2016-09-08 02:01:00.533899 I | etcdmain: etcd Version: 2.2.1
2016-09-08 02:01:00.533949 I | etcdmain: Git SHA: 75f8282
2016-09-08 02:01:00.533953 I | etcdmain: Go Version: go1.5.1
2016-09-08 02:01:00.533957 I | etcdmain: Go OS/Arch: linux/amd64
2016-09-08 02:01:00.533964 I | etcdmain: setting maximum number of CPUs to 1, total number of available CPUs is 1
2016-09-08 02:01:00.534010 N | etcdmain: the server is already initialized as member before, starting as etcd member...
2016-09-08 02:01:00.534068 I | etcdmain: listening for peers on http://127.0.0.1:2380
2016-09-08 02:01:00.534088 I | etcdmain: listening for client requests on http://127.0.0.1:4001
2016-09-08 02:01:00.535059 W | snap: skipped unexpected non snapshot file db
2016-09-08 02:01:00.535071 I | etcdserver: name = default
2016-09-08 02:01:00.535075 I | etcdserver: data dir = /var/etcd/data
2016-09-08 02:01:00.535078 I | etcdserver: member dir = /var/etcd/data/member
2016-09-08 02:01:00.535082 I | etcdserver: heartbeat = 100ms
2016-09-08 02:01:00.535086 I | etcdserver: election = 1000ms
2016-09-08 02:01:00.535089 I | etcdserver: snapshot count = 10000
2016-09-08 02:01:00.535097 I | etcdserver: advertise client URLs = http://127.0.0.1:4001
2016-09-08 02:01:00.535127 W | wal: ignored file 0.tmp in wal
panic: runtime error: makeslice: len out of range

goroutine 1 [running]:
github.com/coreos/etcd/wal.(*decoder).decode(0xc820198630, 0xc820186178, 0x0, 0x0)
	/home/vagrant/gopath/src/github.com/coreos/etcd/release/etcd/gopath/src/github.com/coreos/etcd/wal/decoder.go:55 +0x142
github.com/coreos/etcd/wal.(*WAL).ReadAll(0xc820131d40, 0xc820194a60, 0x15, 0x20, 0x2, 0xce2a822cea30bfca, 0x15b0, 0x0, 0x0, 0x0, ...)
	/home/vagrant/gopath/src/github.com/coreos/etcd/release/etcd/gopath/src/github.com/coreos/etcd/wal/wal.go:238 +0x442
github.com/coreos/etcd/etcdserver.readWAL(0xc820194960, 0x19, 0x0, 0x0, 0x0, 0x0, 0x0, 0xc820131d40, 0xc820191750, 0x5, ...)
	/home/vagrant/gopath/src/github.com/coreos/etcd/release/etcd/gopath/src/github.com/coreos/etcd/etcdserver/storage.go:84 +0x227
github.com/coreos/etcd/etcdserver.restartNode(0xc82013a0e0, 0x0, 0x7f74ef1a1028, 0xc8200b1400, 0x0, 0x0, 0x0, 0x0)
	/home/vagrant/gopath/src/github.com/coreos/etcd/release/etcd/gopath/src/github.com/coreos/etcd/etcdserver/raft.go:286 +0xff
github.com/coreos/etcd/etcdserver.NewServer(0xc82013a0e0, 0xc82013a100, 0x0, 0x0)
	/home/vagrant/gopath/src/github.com/coreos/etcd/release/etcd/gopath/src/github.com/coreos/etcd/etcdserver/server.go:288 +0x322d
github.com/coreos/etcd/etcdmain.startEtcd(0xc8200fa6c0, 0x0, 0x0, 0x0)
	/home/vagrant/gopath/src/github.com/coreos/etcd/release/etcd/gopath/src/github.com/coreos/etcd/etcdmain/etcd.go:287 +0x1a92
github.com/coreos/etcd/etcdmain.Main()
	/home/vagrant/gopath/src/github.com/coreos/etcd/release/etcd/gopath/src/github.com/coreos/etcd/etcdmain/etcd.go:117 +0x1e26
main.main()
	/home/vagrant/gopath/src/github.com/coreos/etcd/release/etcd/gopath/src/github.com/coreos/etcd/main.go:29 +0x14

goroutine 12 [chan receive]:
github.com/coreos/etcd/Godeps/_workspace/src/github.com/golang/glog.(*loggingT).flushDaemon(0x11b4240)
	/home/vagrant/gopath/src/github.com/coreos/etcd/release/etcd/gopath/src/github.com/coreos/etcd/Godeps/_workspace/src/github.com/golang/glog/glog.go:879 +0x67
created by github.com/coreos/etcd/Godeps/_workspace/src/github.com/golang/glog.init.1
	/home/vagrant/gopath/src/github.com/coreos/etcd/release/etcd/gopath/src/github.com/coreos/etcd/Godeps/_workspace/src/github.com/golang/glog/glog.go:410 +0x297

goroutine 32 [syscall]:
os/signal.loop()
	/usr/local/go/src/os/signal/signal_unix.go:22 +0x18
created by os/signal.init.1
	/usr/local/go/src/os/signal/signal_unix.go:28 +0x37
2016-09-08 02:01:02.244576 I | etcdmain: etcd Version: 2.2.1
2016-09-08 02:01:02.249642 I | etcdmain: Git SHA: 75f8282
2016-09-08 02:01:02.249658 I | etcdmain: Go Version: go1.5.1
2016-09-08 02:01:02.249662 I | etcdmain: Go OS/Arch: linux/amd64
2016-09-08 02:01:02.249670 I | etcdmain: setting maximum number of CPUs to 1, total number of available CPUs is 1
```

@kubernetes/sig-scalability @wojtek-t 
we probably at least want to triage for 1.4 so marking as p0.

### PR DESCRIPTION
**Pull request by lavalamp**
This will switch to etcd 2.3.7 for release 1.4, to resolve issues rolling back from 1.4 to 1.3 (while preventing those same issues rolling back to 1.4.0 from a release including etcd 3.0.x).

Fixes #32253.

See https://github.com/kubernetes/kubernetes/issues/32253#issuecomment-245735209 for etcd roadmap.

---

This change is [Reviewable](https://reviewable.kubernetes.io/reviews/kubernetes/kubernetes/32321)

### CODE REVIEW

**Comment by lavalamp** [comment]
fyi @wojtek-t

**Comment by timothysc** [comment]
Generally lgtm, but you may want to chat with @ixdy regarding some other container version matching issues they are having.

**Comment by lavalamp** [comment]
I think @ixdy didn't actually make that change, but we can ask him. That'd go in another repo, anyway.

File: build/build-image/cross/Dockerfile [hunk 1]
```diff
@@ -72,7 +72,7 @@ RUN mkdir $TMPDIR \
             github.com/jteeuwen/go-bindata/go-bindata
 
 # Download and symlink etcd. We need this for our integration tests.
-RUN export ETCD_VERSION=v3.0.4; \
+RUN export ETCD_VERSION=v2.3.7; \
```
Comment (ixdy) [inline build/build-image/cross/Dockerfile]: #32245 is updating the cross-build image to v1.6.3-4 with etcd 3.0.4. You'll probably need to update cross/VERSION to 1.6.3-5 and push a new version.

Comment (lavalamp) [inline build/build-image/cross/Dockerfile]: Working on it.

Comment (lavalamp) [inline build/build-image/cross/Dockerfile]: 1.6.3-5 is pushed.

**Comment by xiang90** [comment]
lgtm

**Comment by lavalamp** [comment]
Applying label.

**Comment by lavalamp** [comment]
It looks like this broke @wojtek-t's migration logic.

**Comment by wojtek-t** [comment]
@lavalamp - this is kind of expected, because now we are using 2.3.7 version of etcd, which obviously doesn't support etcd v3 api.
The way to fix it is to modify test-update-storage-objects.sh to:
- change STORAGE_BACKEND_ETCD3 TO STORAGE_BACKEND_ETCD2 in all places using these variables
- comment about "Step2" phase completely.
- remove "ETCDCTL_API=3" in step 3

This should be enough.

**Comment by wojtek-t** [comment]
OK - I patched this PR into my own, where I fixed the test.

Closing in favor of https://github.com/kubernetes/kubernetes/pull/32359

**Comment by lavalamp** [comment]
Thanks!

**Pull request closed without merging**